platform/linux-generic/odp_comp.c

```diff
@@ -68,42 +514,149 @@ odp_comp_hash_alg_capability(odp_comp_hash_alg_t hash,
 {
 	(void)capa;
 	switch (hash) {
+	case ODP_COMP_HASH_ALG_NONE:
+		capa[0].digest_len = 0;
+		return 0;
 	default:
 		return -1;
 	}
 	return -1;
 }
 
+static int _odp_comp_single(odp_packet_t pkt_in, odp_packet_t pkt_out,
+			    const odp_comp_packet_op_param_t *param)
+{
+	odp_comp_generic_session_t *session;
+	odp_comp_packet_result_t *result;
+	int rc;
+
+	session = to_gen_session(param->session);
+	ODP_ASSERT(session);
+	ODP_ASSERT(pkt_in != ODP_PACKET_INVALID);
+	ODP_ASSERT(pkt_out != ODP_PACKET_INVALID);
+
+	result = get_op_result_from_packet(pkt_out);
+	ODP_ASSERT(result);
+
+	result->pkt_in = pkt_in;
+	result->output_data_range.offset = param->out_data_range.offset;
+	result->output_data_range.length = 0;
+
+	packet_subtype_set(pkt_out, ODP_EVENT_PACKET_COMP);
+
+	rc = session->comp.func(pkt_in, pkt_out, param, session);
+	if (rc < 0)
+		return rc;
+
+	return 0;
+}
+
 int odp_comp_op(const odp_packet_t pkt_in[], odp_packet_t pkt_out[],
 		int num_pkt, const odp_comp_packet_op_param_t param[])
 {
-	(void)pkt_in;
-	(void)pkt_out;
-	(void)num_pkt;
-	(void)param;
+	int i;
+	int rc;
 
-	return -1;
+	for (i = 0; i < num_pkt; i++) {
+		rc = _odp_comp_single(pkt_in[i], pkt_out[i], &param[i]);
+		if (rc < 0)
+			break;
+	}
+
+	return i;
 }
 
 int odp_comp_op_enq(const odp_packet_t pkt_in[], odp_packet_t pkt_out[],
 		    int num_pkt, const odp_comp_packet_op_param_t param[])
 {
-	(void)pkt_in;
-	(void)pkt_out;
-	(void)num_pkt;
-	(void)param;
+	int i;
+	int rc;
 
-	return -1;
+	for (i = 0; i < num_pkt; i++) {
+		odp_event_t event;
+		odp_comp_generic_session_t *session;
+
+		rc = _odp_comp_single(pkt_in[i], pkt_out[i], &param[i]);
+		if (rc < 0)
+			break;
+
+		event = odp_packet_to_event(pkt_out[i]);
+		session = to_gen_session(param[i].session);
+		if (odp_queue_enq(session->params.compl_queue, event)) {
+			odp_event_free(event);
+			break;
+		}
+	}
+
+	return i;
 }
 
 int odp_comp_result(odp_comp_packet_result_t *result,
 		    odp_packet_t packet)
 {
-	(void)result;
-	(void)packet;
+	odp_comp_packet_result_t *op_result;
+
+	ODP_ASSERT(odp_event_subtype(odp_packet_to_event(packet))
+		   == ODP_EVENT_PACKET_COMP);
+
+	op_result = get_op_result_from_packet(packet);
+	ODP_DBG("Copy operational result back\n");
+	memcpy(result, op_result, sizeof(*result));
 	return 0;
 }
 
+int _odp_comp_init_global(void)
+{
+	size_t mem_size;
+	odp_shm_t shm;
+	int idx;
+
+	/* Calculate the memory size we need */
+	mem_size = sizeof(*global);
+
+	/* Allocate our globally shared memory */
+	shm = odp_shm_reserve("comp_pool", mem_size, ODP_CACHE_LINE_SIZE, 0);
+
+	global = odp_shm_addr(shm);
+
+	/* Clear it out */
+	memset(global, 0, mem_size);
```

Comment:
It would be good to check if shm reserve succeeded before using the memory. Output error message and return, if reserve fails.